core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java

```diff
@@ -382,12 +383,22 @@ protected boolean booleanEquality(SqlCallBinding binding,
     return false;
   }
 
+  @Deprecated @Override public boolean caseWhenCoercion(SqlCallBinding callBinding) {
+    return caseOrEquivalentCoercion(callBinding);
+  }
+
   /**
-   * CASE and COALESCE type coercion, collect all the branches types including then
+   * CASE WHEN and COALESCE type coercion, collect all the branches types including then
    * operands and else operands to find a common type, then cast the operands to the common type
    * when needed.
    */
-  @Override public boolean caseWhenCoercion(SqlCallBinding callBinding) {
+  @Override public boolean caseOrEquivalentCoercion(SqlCallBinding callBinding) {
+    if (!(callBinding.getCall() instanceof SqlCase)
```

Comment:
Do we need to judge the SqlCase type here?

Reply:
Maybe I didn't describe it clearly. Is `callBinding.getCall().getKind() == SqlKind.COALESCE` not enough to determine whether it is COALESC?

Reply:
it is.
I think looking at the kind is more efficient too than calling instanceof
I favor positive checks == instead of !=, the meaning is unambiguous when you refactor the code.